rust-runtime/aws-smithy-checksums/src/lib.rs

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Checksum calculation and verification callbacks
+
+use aws_smithy_http::callback::BodyCallback;
+use aws_smithy_types::base64;
+
+use http::header::{HeaderMap, HeaderName, HeaderValue};
+use sha1::Digest;
+use std::io::Write;
+
+const CRC_32_NAME: &str = "x-amz-checksum-crc32";
+const CRC_32_C_NAME: &str = "x-amz-checksum-crc32c";
+const SHA_1_NAME: &str = "x-amz-checksum-sha1";
+const SHA_256_NAME: &str = "x-amz-checksum-sha256";
+
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+#[derive(Debug, Default)]
+struct Crc32callback {
+    hasher: crc32fast::Hasher,
+}
+
+impl Crc32callback {
+    fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
+        self.hasher.update(bytes);
+
+        Ok(())
+    }
+
+    fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> {
+        let mut header_map = HeaderMap::new();
+        let key = HeaderName::from_static(CRC_32_NAME);
+        // We clone the hasher because `Hasher::finalize` consumes `self`
+        let hash = self.hasher.clone().finalize();
+        let value = HeaderValue::from_str(&base64::encode(u32::to_be_bytes(hash)))
+            .expect("base64 will always produce valid header values from checksums");
+
+        header_map.insert(key, value);
+
+        Ok(Some(header_map))
+    }
+}
+
+impl BodyCallback for Crc32callback {
+    fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
+        self.update(bytes)
+    }
```

Comment:
On all of these: why implement `update` and `trailers` in a separate impl block?

Reply:
The original idea would be that it allows you to call these functions without needing to use the trait. This would happen within the "mutate request" section for non-streaming requests. For streaming requests they'd all be boxed trait objects.

I'm fine with making them only usable behind the trait if you like. Also, I don't think I've mentioned that these callbacks are going to be marked `pub` when I do the codegen half of things. I made them private to start so people wouldn't touch them.